### PR TITLE
Do not save duplicate tags

### LINKED ISSUE #359: Ensure there are no duplicate tags
As is stands, multiple tags with the same name can be created. Make it so that this is not the case. For example, if a tag called "dystopian" exists, another tag with the same name should not be allowed to be created.

This change should affect the `TagService`. An accompanying test is also required.

The check for a tag with the same name should be case insensitive (i.e. 'Dystopian' should not be permitted if 'dystopian' already exists)

**Additional context**

Branch from `0.2.0` and create a branch to go back into it

### PR DESCRIPTION
**Pull request by VikramSitaram**
## Summary of change

- Add duplicate check to save method inside `TagService.java`
- Add test inside `TagServiceTest.java` 
- Add CascadeType.REMOVE to `books` field inside the `Tag` entity

## Related issue

Closes #359 

## Pull request checklist

Please keep this checklist in & ensure you have done the following:

- [X] Read, understood and adhered to our [contributing document](https://github.com/knjk04/book-project/blob/master/CONTRIBUTING.md).
  - [X] Ensure that you were first assigned to a relevant issue before creating this pull request
  - [X] Ensure code changes pass all tests

- [X] Read, understood and adhered to our [style guide](https://github.com/knjk04/book-project/blob/master/STYLEGUIDE.md). A lot of our code reviews are spent on ensuring compliance with our style guide, so it would save a lot of time if this was adhered to from the outset. 

- [X] Filled in the summary, context (if applicable) and related issue section. Replace the square brackets and its placeholder content with your contents. For an example, see any merged in pull request
  - [X] Included a screenshot(s) if a UI change was involved (it may look different on a reviewer's device)

- [X] Created a branch that has a descriptive name (what your branch is for in a few words and includes the issue number at the end, e.g. `test-reading-goal-123`

- [X] Set this pull request to 'draft' if you are still working on it

- [X] Resolved any merge conflicts

For any of the optional checkboxes (e.g. the screenshots one), still check it if it does not apply.

If in doubt, get in touch with us via our [Slack workspace](https://join.slack.com/t/teambookproject/shared_invite/zt-jcijyenp-JiKFGBv62FIPoFnvOW6Ubg)

### CODE REVIEW

File: backend/src/main/java/com/karankumar/bookproject/backend/model/Tag.java [hunk 1]
```diff
@@ -50,7 +51,7 @@
     @Column(unique = true)
     private String name;
 
-    @ManyToMany(fetch = FetchType.EAGER, mappedBy = "tags")
+    @ManyToMany(fetch = FetchType.EAGER, cascade = CascadeType.REMOVE, mappedBy = "tags")
```
Comment (knjk04) [inline backend/src/main/java/com/karankumar/bookproject/backend/model/Tag.java:54]: Could we remove the new cascade type? We don't want to remove a book if a tag is deleted

File: backend/src/main/java/com/karankumar/bookproject/backend/service/TagService.java [hunk 2]
```diff
@@ -38,7 +38,19 @@ public Tag findById(Long id) {
     }
 
     public void save(@NonNull Tag tag) {
-        tagRepository.save(tag);
+        boolean hasDuplicate = false;
```
Comment (knjk04) [inline backend/src/main/java/com/karankumar/bookproject/backend/service/TagService.java:41]: Thanks for this! This works, but could you instead create a derived query that searches for a tag by name (that's case insensitive)? That would be useful in other places to find books that belong to a particular tag. 

I appreciate I didn't mention this before, but it just came to mind when looking at this method's implementation

**Review by knjk04 — CHANGES_REQUESTED** [review]
Thanks for working on this! I've left some feedback for you to address.

I had a look into the failing test and I wasn't able to fix it just yet, so I'll create a separate issue for that once this is merged in. This looks good, otherwise, though

File: backend/src/main/java/com/karankumar/bookproject/backend/model/Tag.java (hunk 1 above)
Comment (VikramSitaram) [reply backend/src/main/java/com/karankumar/bookproject/backend/model/Tag.java]: Removed but this introduces a set of new test errors (which was the orif=ginal reason I added the CasacdeType.REMOVE). The test errors (not failures) can be seen here --> https://pastebin.com/yC60U37C.

File: backend/src/main/java/com/karankumar/bookproject/backend/service/TagService.java (hunk 2 above)
Comment (VikramSitaram) [reply backend/src/main/java/com/karankumar/bookproject/backend/service/TagService.java]: Fair enough, I added a findByName() method to TagRepository and this is referenced inside TagService. The test that I wrote for this seems to work as well.

**Review by knjk04 — APPROVED** [review]
Thanks for making the changes! I've also made some small adjustments.

Could you please remove the `backend/mvnw` change? As this file is now empty, it won't work on Linux & macOS

**Comment by VikramSitaram** [comment]
> Thanks for working on this! I've left some feedback for you to address.
> 
> I had a look into the failing test and I wasn't able to fix it just yet, so I'll create a separate issue for that once this is merged in. This looks good, otherwise, though

Sounds good, thanks!

**Comment by VikramSitaram** [comment]
> Thanks for making the changes! I've also made some small adjustments.
> 
> Could you please remove the `backend/mvnw` change? As this file is now empty, it won't work on Linux & macOS

Thanks for the changes, looks much more concise now. I've reverted the changes to mvnw as well.

**Comment by knjk04** [comment]
@VikramSitaram Thanks for working on this!

**Pull request merged**